README.md
# OBS Mic Check

Plays a audio alarm (media source) when [OBS Studio](https://obsproject.com/) Audio mute state does not match with displayed sources (BRB, Starting Soon, etc)

## Requirements

- [OBS Studio](https://obsproject.com/)

## Installation

There are three scripts which cooperate:

- obs-script-mic-check.lua: Core monitoring and script dialog settings, The other scripts expect to interact with this.
- obs-script-mic-check-source-settings-filter.lua: Provides a filter used to do per-source rules. Technically optional, though source-aware filters are a core feature.
- obs-script-mic-check-status-monitor.lua: Provides a custom source that displays which sources have attached rules filters, and the status of the alarm triggers. Intended primarily for window projector. Optional.

## Usage

Add a media source for the alarm. A suitable sound file is provided with the script. Open Advanced Audio Properties for the source and change Audio Monitoring to Monitor Only (mute output).

Add a copy of the alarm source to every scene where you want to hear it.

Attach rules to video sources ("BRB", "Starting Soon", etc) using the "Mic Check Settings" filter. (Right-click on a source and select filters.) The first active video source with attached settings will be used to trigger alarms instead of the defaults.

If no such video source is active, then the default rules on the script dialog will be used.

## Credits

Alert sounds: [`pup_alert.mp3`](https://freesound.org/people/willy_ineedthatapp_com/sounds/167337/) by `willy_ineedthatapp_com`


CHANGELOG.md
# OBS Mic Check Changelog

## 1.0.1

- Status monitor was trying to use source active status, but not actually reading the source active status
- Status monitor was not considering inactive sources as muted
- Rule evaluation was not considering inactive sources as muted


lib/obs-script-mic-check-common.lua
local obs = obslua

alarm_active = false
trigger_active = false
trigger_time = os.time()
trigger_timeout = 0
audio_sources = {}
video_sources = {}
default_rule = {
	timeout = 5,
	operator = "all",
	audio_states = {
		["Mic/Aux"] = "muted"
	}
}
source_rules = {}

function dump_rule(rule)
	local items = {}
	for source,state in pairs(rule.audio_states) do
		table.insert(items, source .. "=" .. state)
	end
	script_log(rule.operator .. "(" .. table.concat(items, ",") .. ")")
end

function serialize_rule(rule)
	local data = obs.obs_data_create()
	local states = obs.obs_data_create()

	obs.obs_data_set_int(data, 'timeout', rule.timeout)
	obs.obs_data_set_string(data, 'operator', rule.operator)
	if rule.name then
		obs.obs_data_set_string(data, 'name', rule.name)
	end
	obs.obs_data_set_obj(data, 'audio_states', states)

	for name,state in pairs(rule.audio_states) do
		obs.obs_data_set_string(states, name, state)
	end

	local json = obs.obs_data_get_json(data)

	obs.obs_data_release(data)
	obs.obs_data_release(states)

	return json
end

function deserialize_rule(json)
	local data = obs.obs_data_create_from_json(json)
	local rule = {audio_states = {}}

	rule.timeout = obs.obs_data_get_int(data, 'timeout')
	rule.operator = obs.obs_data_get_string(data, 'operator')
	local name = obs.obs_data_get_string(data, 'name')
	if name ~= '' then
		rule.name = name
	end

	local states = obs.obs_data_get_obj(data, 'audio_states')

	for sourcename,_ in pairs(audio_sources) do
		local state = obs.obs_data_get_string(states, sourcename)
		if state ~= '' then
			rule.audio_states[sourcename] = state
		end
	end

	obs.obs_data_release(data)
	obs.obs_data_release(states)

	return rule
end

function update_source_rule(calldata)
	script_log('receive source')
	local id = obs.calldata_int(calldata, 'id')
	local json = obs.calldata_string(calldata, 'rule_json')
	local rule = deserialize_rule(json)
	source_rules[id] = rule
	dump_rule(rule)
end

function bootstrap_rule_settings(rule, settings)
	if rule == nil then
		script_log("bootstrap_rule_settings no rule")
	end
	if settings == nil then
		script_log("bootstrap_rule_settings no settings")
	end
	rule.timeout = obs.obs_data_get_int(settings, "timeout")
	rule.operator = obs.obs_data_get_string(settings, "operator")

	rule.audio_states = {}
	local index = 1
	while index < 200 do
		local name = obs.obs_data_get_string(settings, "audio" .. index)
		if name == nil or name == "" then
			break
		end
		index = index + 1

		local state = obs.obs_data_get_string(settings, name)
		if state == audio_status(true) or state == audio_status(false) then
			rule.audio_states[name] = state
			if audio_sources[name] == nil then
				audio_sources[name] = {
					name = name,
					status = audio_status(false),
					active = video_status(false),
					flags = 0,
					in_current_scene = false,
				}
			end
		end
	end

	dump_rule(rule)
end

function update_rule_settings(rule, settings)
	if rule == nil then
		script_log("update_rule_settings no rule")
	end
	if settings == nil then
		script_log("update_rule_settings no settings")
	end
	rule.timeout = obs.obs_data_get_int(settings, "timeout")
	rule.operator = obs.obs_data_get_string(settings, "operator")
	rule.audio_states = {}
	local index = 1
	for _,source in pairs(audio_sources) do
		local state = obs.obs_data_get_string(settings, source.name)
		if state == audio_status(true) or state == audio_status(false) then
			rule.audio_states[source.name] = state
			obs.obs_data_set_string(settings, "audio" .. index, source.name)
			index = index + 1
		end
	end
	obs.obs_data_erase(settings, "audio" .. index)
	dump_rule(rule)
end

function audio_default_settings(settings)
	obs.obs_data_set_default_int(settings, "timeout", 5)
	obs.obs_data_set_default_string(settings, "operator", "any")
	for _,source in pairs(audio_sources) do
		obs.obs_data_set_default_string(settings, source.name, "disabled")
	end
end

function run_rule(rule)
	for name,status in pairs(rule.audio_states) do
		local cache = audio_sources[name]
		if cache then
			if (cache.active == "active" and cache.status == status)
					or (cache.active == "hidden" and status == "muted") then
				if rule.operator == "any" then
					return true, rule.timeout
				end
			else
				if rule.operator == "all" then
					return false, rule.timeout
				end
			end
		end
	end

	if rule.operator == "any" then
		return false, rule.timeout
	else
		return true, rule.timeout
	end
end

function audio_status(muted)
	if muted then
		return "muted"
	else
		return "live"
	end
end

function video_status(active)
	if active then
		return "active"
	else
		return "hidden"
	end
end

local function source_active(calldata, active_flag)
	local source = obs.calldata_source(calldata, "source")
	local active = video_status(active_flag)
	local name = obs.obs_source_get_name(source)
	script_log(name .. " " .. active .. " " .. obs.obs_source_get_id(source))
	local cache = video_sources[name]
	if cache then
		cache.active = active
	end
end

function source_activate(calldata)
	source_active(calldata, true)
end

function source_deactivate(calldata)
	source_active(calldata, false)
end

-- A function named script_description returns the description shown to
-- the user
local description = [[Support script for obs-script-mic-check, should not be loaded directly into OBS.]]
function script_description()
	return description
end


obs-script-mic-check-status-monitor.lua
local obs = obslua
local bit = require("bit")
local os = require("os")

function script_log(message) -- luacheck: no unused args
	-- unreachable code
	-- luacheck: push ignore
	if false then
		obs.script_log(obs.LOG_INFO, message)
	end
	-- luacheck: pop
end

local status_margin = 10
local status_width = 640
local status_height = 500
local status_font_size = 40
local status_indent = 100

local text_white = 0xffffffff
local text_yellow = 0xff44ffff
local text_gray = 0xffaaaaaa

dofile(script_path() .. "lib/obs-script-mic-check-common.lua")

local function source_mute(calldata)
	script_log("receive mute")
	local source = obs.calldata_source(calldata, "source")
	local status = audio_status(obs.obs_source_muted(source))
	local active = video_status(obs.obs_source_active(source))
	local name = obs.obs_source_get_name(source)
	script_log(name .. " " .. status .. " " .. obs.obs_source_get_id(source))
	local cache = audio_sources[name]
	if cache then
		cache.active = active
		cache.status = status
	else
		audio_sources[name] = {
			name = name,
			active = active,
			status = status,
		}
	end
end

local update_default_rule = function(calldata)
	script_log('receive default')
	local json = obs.calldata_string(calldata, 'rule_json')
	default_rule = deserialize_rule(json)
	dump_rule(default_rule)
end

local video_source_status = function(calldata)
	local name = obs.calldata_string(calldata, 'name')
	local active = obs.calldata_bool(calldata, 'active')
	local in_current_scene = obs.calldata_bool(calldata, 'in_current_scene')
	script_log('receive video status ' .. name .. ' ' .. video_status(active))
	video_sources[name] = {
		name = name,
		active = video_status(active),
		in_current_scene = in_current_scene,
	}
end

local trigger_activate = function(calldata)
	trigger_active = true
	trigger_time = obs.calldata_int(calldata, 'time')
	trigger_timeout = obs.calldata_int(calldata, 'timeout')
end

local alarm_activate = function()
	alarm_active = true
end

local alarm_reset = function()
	alarm_active = false
	trigger_active = false
end

local bootstrap = function()
	local sh = obs.obs_get_signal_handler()
	obs.signal_handler_signal(sh, "lua_mic_check_request_rules", nil)
	obs.remove_current_callback()
end

-- A function named script_description returns the description shown to
-- the user
-- luacheck: push no max line length
local description = [[OBS Source which visually displays mic check alarm status.

Add to your sources, it may be most usefull hidden and used with window projector so it is not visible on stream.

Displays the default rule and a section for each source with a attached settings filter. Only one set of rules is active at any given time, the associated rule set will have a colored bar.

Yellow audio sources indicated they are currently in violation of the rule.

A countup timer will be shown when a rule is triggered and the alarm will be activated soon. The entire source is red when the alarm is active.

Without obs-script-mic-check.lua functionality will be limited to list the sources with attached settings filters.]]
-- luacheck: pop
function script_description()
	return description
end

-- a function named script_load will be called on startup
function script_load(settings) -- luacheck: no unused args
	script_log("script status load")

	local sh = obs.obs_get_signal_handler()
	-- signals received
	obs.signal_handler_connect(sh, "source_activate", source_activate)
	obs.signal_handler_connect(sh, "source_deactivate", source_deactivate)
	obs.signal_handler_add(sh, "void lua_mic_check_source_mute(ptr source)")
	obs.signal_handler_connect(sh, "lua_mic_check_source_mute", source_mute)
	obs.signal_handler_add(sh, "void lua_mic_check_default_rule(string rule_json)")
	obs.signal_handler_connect(sh, "lua_mic_check_default_rule", update_default_rule)
	obs.signal_handler_add(sh, "void lua_mic_check_source_rule(int id, string rule_json)")
	obs.signal_handler_connect(sh, "lua_mic_check_source_rule", update_source_rule)
	obs.signal_handler_add(sh, "void lua_mic_check_video_source_status(string name, bool active, bool in_current_scene)")
	obs.signal_handler_connect(sh, "lua_mic_check_video_source_status", video_source_status)
	obs.signal_handler_add(sh, "void lua_mic_check_trigger_activate(int time, int timeout)")
	obs.signal_handler_connect(sh, "lua_mic_check_trigger_activate", trigger_activate)
	obs.signal_handler_add(sh, "void lua_mic_check_alarm_activate()")
	obs.signal_handler_connect(sh, "lua_mic_check_alarm_activate", alarm_activate)
	obs.signal_handler_add(sh, "void lua_mic_check_alarm_reset()")
	obs.signal_handler_connect(sh, "lua_mic_check_alarm_reset", alarm_reset)

	-- signals sent
	obs.signal_handler_add(sh, "void lua_mic_check_request_audio_sources()")
	obs.signal_handler_add(sh, "void lua_mic_check_request_rules()")

	obs.signal_handler_signal(sh, "lua_mic_check_request_audio_sources", nil)

	obs.timer_add(bootstrap, 1000)
end

local create_label = function(name, size, color)
	local settings = obs.obs_data_create()
	local font = obs.obs_data_create()

	obs.obs_data_set_string(font, "face", "Arial")
	obs.obs_data_set_int(font, "size", size * 0.90) -- freetype is larger than gdiplus

	obs.obs_data_set_obj(settings, "font", font)
	obs.obs_data_set_string(settings, "text", " " .. name .. " ")
	obs.obs_data_set_int(settings, "color", color) -- gdiplus
	obs.obs_data_set_int(settings, "color1", color) -- freetype
	obs.obs_data_set_int(settings, "color2", color) -- freetype

	obs.obs_enter_graphics();
	--local source = obs.obs_source_create_private("text_gdiplus", name .. "-label", settings)
	local source = obs.obs_source_create_private("text_ft2_source", name .. "-label", settings)
	obs.obs_leave_graphics();
	obs.obs_data_release(font)
	obs.obs_data_release(settings)

	return source
end

local set_label_text = function(source, text)
	local settings = obs.obs_source_get_settings(source)
	obs.obs_data_set_string(settings, "text", " " .. text .. " ")
	obs.obs_source_update(source, settings)
	obs.obs_data_release(settings)
end

local draw_label = function(data, key, name, size, color)
	if data.labels[key] == nil then
		script_log("create " .. key)
		data.labels[key] = create_label(name, size, color)
	end
	obs.obs_source_video_render(data.labels[key])
end

local data_paths = {
	"../../data/",
	"../data/",
}

local function image_source_load(image, file)
	for _,data in ipairs(data_paths) do
		obs.obs_enter_graphics();
		obs.gs_image_file_free(image);
		obs.obs_leave_graphics();

		obs.gs_image_file_init(image, data .. file);

		obs.obs_enter_graphics();
		obs.gs_image_file_init_texture(image);
		obs.obs_leave_graphics();

		if image.loaded then
			break
		end
	end

	if not image.loaded then
		print("failed to load texture " .. file);
	end
end

local source_def = {}
source_def.id = "lua_mic_check_status_source"
source_def.output_flags = bit.bor(obs.OBS_SOURCE_VIDEO, obs.OBS_SOURCE_CUSTOM_DRAW)

source_def.get_name = function()
	return "Mic Check Status Monitor"
end

source_def.create = function(source, settings) -- luacheck: no unused args
	script_log("source status create")
	local data = {
		labels = {},
		live_image = obs.gs_image_file(),
		mute_image = obs.gs_image_file(),
		height = status_height,
	}

	image_source_load(data.live_image, "obs-studio/themes/Dark/unmute.png")
	image_source_load(data.mute_image, "obs-studio/themes/Dark/mute.png")

	return data
end

source_def.destroy = function(data)
	if data == nil then
		return
	end

	obs.obs_enter_graphics();

	for key,_ in pairs(data.labels) do
		-- frequently deadlocks OBS when reloading (unloading in general?)
		--obs.obs_source_release(label)
		data.labels[key] = nil
	end

	obs.gs_image_file_free(data.live_image)
	obs.gs_image_file_free(data.mute_image)

	obs.obs_leave_graphics()
end

local function status_item(data, title, rule, controlling)
	local effect_default = obs.obs_get_base_effect(obs.OBS_EFFECT_DEFAULT)
	local effect_solid = obs.obs_get_base_effect(obs.OBS_EFFECT_SOLID)
	local color_param = obs.gs_effect_get_param_by_name(effect_solid, "color");

	local violation = run_rule(rule)
	local source = video_sources[rule.name]
	local in_current_scene = title == 'Default' or (source ~= nil and source.in_current_scene)
	local height = 0

	if controlling then
		if violation then
			obs.gs_effect_set_color(color_param, 0xff888800)
		else
			obs.gs_effect_set_color(color_param, 0xff008800)
		end
	elseif in_current_scene then
		obs.gs_effect_set_color(color_param, 0xff666666)
	else
		obs.gs_effect_set_color(color_param, 0xff777777)
	end
	while obs.gs_effect_loop(effect_solid, "Solid") do
		obs.gs_draw_sprite(nil, 0, status_width - status_margin*2, status_font_size)
	end
	local heading = title.."-white"
	local text_color = text_white
	if not in_current_scene then
		heading = title.."-gray"
		text_color = text_gray
	end
	if title then
		draw_label(data, heading, title, status_font_size, text_color)
	end
	obs.gs_matrix_translate3f(0, status_font_size, 0)
	height = height + status_font_size
	if rule.operator == 'any' or rule.operator == 'all' then
		if violation then
			draw_label(data, rule.operator..'-yellow', rule.operator, status_font_size, text_yellow)
		else
			draw_label(data, rule.operator..'-white', rule.operator, status_font_size, text_white)
		end
	end
	obs.gs_matrix_push()
	obs.gs_matrix_translate3f(status_indent, 0, 0)
	local items = 0
	for name,status in pairs(rule.audio_states) do
		local image = data.live_image
		if status == 'muted' then
			image = data.mute_image
		end

		obs.gs_matrix_push()
		obs.gs_matrix_translate3f(0.1 * status_font_size, 0.1 * status_font_size, 0)
		obs.gs_matrix_scale3f(0.8 * status_font_size / image.cy, 0.8 * status_font_size / image.cy, 0)
		while obs.gs_effect_loop(effect_default, "Draw") do
			obs.obs_source_draw(image.texture, 0, 0, image.cx, image.cy, false);
		end
		obs.gs_matrix_pop()

		obs.gs_matrix_push()
		obs.gs_matrix_translate3f(50, 0, 0)
		if audio_sources[name] == nil then
			draw_label(data, name..'-gray', name, status_font_size, text_gray)
		elseif (audio_sources[name].active == "active"
			and audio_sources[name].status == status)
			or (audio_sources[name].active == "hidden"
			and status == "muted") then
			draw_label(data, name..'-yellow', name, status_font_size, text_yellow)
		else
			draw_label(data, name..'-white', name, status_font_size, text_white)
		end
		obs.gs_matrix_pop()

		obs.gs_matrix_translate3f(0, status_font_size, 0)
		items = items + 1
	end
	obs.gs_matrix_pop()
	local offset = status_font_size * math.max(1, items)
	obs.gs_matrix_translate3f(0, offset, 0)
	height = height + offset
	return height
end

source_def.video_render = function(data, effect) -- luacheck: no unused args
	if data == nil then
		return
	end

	obs.gs_blend_state_push()
	obs.gs_reset_blend_state()

	local effect_solid = obs.obs_get_base_effect(obs.OBS_EFFECT_SOLID)
	local color_param = obs.gs_effect_get_param_by_name(effect_solid, "color");

	if alarm_active then
		obs.gs_effect_set_color(color_param, 0xffaa4444)
	else
		obs.gs_effect_set_color(color_param, 0xff444444)
	end
	while obs.gs_effect_loop(effect_solid, "Solid") do
		obs.gs_draw_sprite(nil, 0, status_width, data.height)
	end

	obs.gs_matrix_push()

	obs.gs_matrix_translate3f(status_margin, status_margin, 0)


	if trigger_active then
		local duration = os.difftime(os.time(), trigger_time)
		local progress = math.min(1, duration / trigger_timeout)
		obs.gs_effect_set_color(color_param, 0xffaa4444)
		while obs.gs_effect_loop(effect_solid, "Solid") do
			obs.gs_draw_sprite(nil, 0, (status_width - status_margin*2) * progress, status_font_size)
		end

		if data.labels['duration'] == nil then
			script_log("create duration")
			data.labels['duration'] = create_label('duration', status_font_size, text_white)
		end

		set_label_text(data.labels['duration'], string.format("%d", duration))
		obs.obs_source_video_render(data.labels['duration'])
	end

	obs.gs_matrix_translate3f(0, status_font_size, 0)
	local height = status_font_size

	local found_first_active = false

	for _,rule in pairs(source_rules) do
		local controlling = false
		local source = video_sources[rule.name]
		local title = rule.name
		if title == nil then
			title = "--"
		end
		if not found_first_active and rule.name then
			if source and source.active == 'active' then
				found_first_active = true
				controlling = true
			end
		end
		if source and source.in_current_scene then
			height = height + status_item(data, title, rule, controlling)
		end
	end

	for _,rule in pairs(source_rules) do
		local controlling = false
		local source = video_sources[rule.name]
		local title = rule.name
		if title == nil then
			title = "--"
		end
		if not found_first_active and rule.name then
			if source and source.active == 'active' then
				found_first_active = true
				controlling = true
			end
		end
		if not (source and source.in_current_scene) then
			height = height + status_item(data, title, rule, controlling)
		end
	end
	height = height + status_item(data, "Default", default_rule, not found_first_active)

	data.height = height + status_margin * 2

	obs.gs_matrix_pop()

	obs.gs_blend_state_pop()
end

source_def.get_width = function(data) -- luacheck: no unused args
	return status_width
end

source_def.get_height = function(data)
	return data.height
end

obs.obs_register_source(source_def)


obs-script-mic-check.lua
local obs = obslua
local bit = require("bit")
local os = require("os")

function script_log(message) -- luacheck: no unused args
	-- unreachable code
	-- luacheck: push ignore
	if false then
		obs.script_log(obs.LOG_INFO, message)
	end
	-- luacheck: pop
end

local sample_rate = 1000

local alarm_source = ""

local obs_events = {}

dofile(script_path() .. "lib/obs-script-mic-check-common.lua")

local function enum_sources(callback)
	local sources = obs.obs_enum_sources()
	if sources ~= nil then
		for _,source in ipairs(sources) do
			callback(source)
		end
	end
	obs.source_list_release(sources)
end

local function set_alarm_visible(visible)
	if alarm_source ~= nil then
		local current_source = obs.obs_frontend_get_current_scene()
		local current_scene = obs.obs_scene_from_source(current_source)
		local item = obs.obs_scene_find_source(current_scene, alarm_source)
		if item ~= nil then
			obs.obs_sceneitem_set_visible(item, visible)
		end
		obs.obs_source_release(current_source)
	end
end

local function activate_alarm()
	set_alarm_visible(true)
	obs.remove_current_callback()
end

local function play_alarm()
	script_log("alarm")
	set_alarm_visible(false)
	obs.timer_add(activate_alarm, 500)
end

local function set_alarm(alarming)
	if alarming then
		if not alarm_active then
			play_alarm()
			alarm_active = true
			obs.timer_add(play_alarm, 60*1000)
		end
	else
		if alarm_active then
			alarm_active = false
			set_alarm_visible(false)
			obs.timer_remove(play_alarm)
		end
	end
end

local function trigger_alarm(violation, timeout)
	if violation then
		if trigger_active then
			trigger_timeout = timeout
			if os.difftime(os.time(), trigger_time) > timeout then
				set_alarm(true)
				local sh = obs.obs_get_signal_handler()
				obs.signal_handler_signal(sh, "lua_mic_check_alarm_activate", nil)
			end
		else
			script_log("trigger")
			trigger_active = true
			trigger_time = os.time()
			trigger_timeout = timeout
			local sh = obs.obs_get_signal_handler()
			local calldata = obs.calldata()
			obs.calldata_init(calldata)
			obs.calldata_set_int(calldata, "time", trigger_time)
			obs.calldata_set_int(calldata, "timeout", trigger_timeout)
			obs.signal_handler_signal(sh, "lua_mic_check_trigger_activate", calldata)
			obs.calldata_free(calldata)
		end
	else
		--script_log("no violation")
		if alarm_active or trigger_active then
			local sh = obs.obs_get_signal_handler()
			obs.signal_handler_signal(sh, "lua_mic_check_alarm_reset", nil)
		end
		trigger_active = false
		set_alarm(false)
	end
end

local function check_alarm()
	for _,rule in pairs(source_rules) do
		if rule.name then
			local source = video_sources[rule.name]
			if source and source.active == 'active' then
				trigger_alarm(run_rule(rule), rule.timeout)
				return
			end
		end
	end
	trigger_alarm(run_rule(default_rule), default_rule.timeout)
end

local function examine_source_state(source)
	local current_source = obs.obs_frontend_get_current_scene()
	local current_scene = obs.obs_scene_from_source(current_source)
	local sh = obs.obs_get_signal_handler()
	local name = obs.obs_source_get_name(source)
	local status = audio_status(obs.obs_source_muted(source))
	local active = video_status(obs.obs_source_active(source))
	local flags = obs.obs_source_get_output_flags(source)
	local item = obs.obs_scene_find_source(current_scene, name)
	local in_current_scene = item ~= nil
	--script_log(name .. " " .. active .. " " .. status .. " " .. obs.obs_source_get_id(source) .. " " .. bit.tohex(flags))
	local info = {
		name = name,
		status = status,
		active = active,
		flags = flags,
		in_current_scene = in_current_scene,
	}
	if bit.band(flags, obs.OBS_SOURCE_AUDIO) ~= 0 then
		audio_sources[name] = info
		local audiodata = obs.calldata()
		obs.calldata_init(audiodata)
		obs.calldata_set_ptr(audiodata, "source", source)
		obs.signal_handler_signal(sh, "lua_mic_check_source_mute", audiodata)
		obs.calldata_free(audiodata)
	end
	if bit.band(flags, obs.OBS_SOURCE_VIDEO) ~= 0 then
		video_sources[name] = info
		local videodata = obs.calldata()
		obs.calldata_init(videodata)
		obs.calldata_set_string(videodata, "name", name)
		obs.calldata_set_bool(videodata, "active", obs.obs_source_active(source))
		obs.calldata_set_bool(videodata, "in_current_scene", in_current_scene)
		obs.signal_handler_signal(sh, "lua_mic_check_video_source_status", videodata)
		obs.calldata_free(videodata)
	end
	obs.obs_source_release(current_source)
end

local function examine_source_states()
	enum_sources(examine_source_state)
	check_alarm()
	--return true
end

local function send_default_rule()
	local sh = obs.obs_get_signal_handler()
	local calldata = obs.calldata()
	obs.calldata_init(calldata)
	obs.calldata_set_string(calldata, "rule_json", serialize_rule(default_rule))
	obs.signal_handler_signal(sh, "lua_mic_check_default_rule", calldata)
	obs.calldata_free(calldata)
end

local function process_events()
	for _,event in ipairs(obs_events) do
		script_log("event " .. event.event)
		if event.event == obs.OBS_FRONTEND_EVENT_SCENE_CHANGED then
			examine_source_states()
		elseif event.event == 'request_audio_sources' then
			examine_source_states()
		elseif event.event == 'request_rules' then
			send_default_rule()
		end
	end
	obs_events = {}
end

local function tick()
	process_events()
	check_alarm()
end

local function test_alarm(props, p, set) -- luacheck: no unused args
	play_alarm()
	return true
end

local function source_mute(calldata)
	local source = obs.calldata_source(calldata, "source")
	local status = audio_status(obs.obs_source_muted(source))
	--script_log(obs.obs_source_get_name(source) .. " " .. status .. " " .. obs.obs_source_get_id(source))
	local cache = audio_sources[obs.obs_source_get_name(source)]
	if cache then
		cache.status = status
		check_alarm()
		local sh = obs.obs_get_signal_handler()
		obs.signal_handler_signal(sh, "lua_mic_check_source_mute", calldata)
	end
end

local function hook_source(source)
	if source ~= nil then
		local handler = obs.obs_source_get_signal_handler(source)
		if handler ~= nil then
			local flags = obs.obs_source_get_output_flags(source)
			if bit.band(flags, obs.OBS_SOURCE_AUDIO) ~= 0 then
				obs.signal_handler_connect(handler, "mute", source_mute)
			end
		end
	end
end

local function source_create(calldata)
	local source = obs.calldata_source(calldata, "source")
	hook_source(source)
	examine_source_state(source)
end

local function source_destroy(calldata)
	local source = obs.calldata_source(calldata, "source")
	examine_source_state(source)
end

local function request_audio_sources()
	obs_events[#obs_events+1] = {
		event = 'request_audio_sources'
	}
end

local function request_rules()
	obs_events[#obs_events+1] = {
		event = 'request_rules'
	}
end

local function frontend_event(event, private_data) -- luacheck: no unused args
	script_log("frontend event " .. event)
	if event == obs.OBS_FRONTEND_EVENT_SCENE_CHANGED then
		obs_events[#obs_events+1] = {
			event = event
		}

		-- deadlocks OBS on startup
		--examine_source_states()
	end
end

-- A function named script_description returns the description shown to
-- the user
-- luacheck: push no max line length
local description = [[Play an alarm if mic state not appropriate for sources shown.

Add a media source for the alarm. A suitable sound file is provided with the script. Open Advanced Audio Properties for the source and change Audio Monitoring to Monitor Only (mute output).

Add a copy of the alarm source to every scene where you want to hear it.

Use obs-script-mic-check-source-settings-filter.lua to add rules to video sources.
Use obs-script-mic-check-status-monitor.lua to get a visible source to monitor alarm status.

If no filtered video source is active, then the default rules below will be used.]]
-- luacheck: pop
function script_description()
	return description
end

local function add_audio_rule_properties(props)
	local to = obs.obs_properties_add_int(props,
		"timeout", "For this many seconds", 0, 60 * 60, 5)
	obs.obs_property_set_long_description(to,
		"Alarm if audio is in alarm state for this many seconds.")

	local op = obs.obs_properties_add_list(props,
		"operator", "Operator", obs.OBS_COMBO_TYPE_LIST, obs.OBS_COMBO_FORMAT_STRING)
	obs.obs_property_list_add_string(op, "Any", "any")
	obs.obs_property_list_add_string(op, "All", "all")
	obs.obs_property_set_long_description(op,
		"If multiple audio sources are selected below, how should they be combined.")

	for _,source in pairs(audio_sources) do
		local s = obs.obs_properties_add_list(props,
			source.name, source.name, obs.OBS_COMBO_TYPE_LIST, obs.OBS_COMBO_FORMAT_STRING)
		obs.obs_property_list_add_string(s, "N/A", "disabled")
		obs.obs_property_list_add_string(s, "Mute", audio_status(true))
		obs.obs_property_list_add_string(s, "Live", audio_status(false))
		obs.obs_property_set_long_description(s,
			"Alarm will trigger if this audio source is in the specified state.")
	end
end

-- A function named script_properties defines the properties that the user
-- can change for the entire script module itself
function script_properties()
	script_log("props")

	local props = obs.obs_properties_create()

	local p = obs.obs_properties_add_list(props,
		"alarm_source", "Alarm Media Source",
		obs.OBS_COMBO_TYPE_EDITABLE, obs.OBS_COMBO_FORMAT_STRING)
	enum_sources(function(source)
		local source_id = obs.obs_source_get_id(source)
		if source_id == "ffmpeg_source" then
			local name = obs.obs_source_get_name(source)
			obs.obs_property_list_add_string(p, name, name)
		end
	end)
	obs.obs_property_set_long_description(p,
		"See above for how to create an appropriate media source.")

	local ref = obs.obs_properties_add_button(props,
		"test_alarm", "Test Alarm", test_alarm)
	obs.obs_property_set_long_description(ref,
		"Test activating selected media sources")

	local label = obs.obs_properties_add_text(props,
		"label_default", "Default Checks:", 0)
	obs.obs_property_set_enabled(label, false)

	add_audio_rule_properties(props)

	return props
end

-- A function named script_defaults will be called to set the default settings
function script_defaults(settings)
	script_log("defaults")

	obs.obs_data_set_default_string(settings, "label_default", "Alarm if in this state.")
	obs.obs_data_set_default_string(settings, "alarm_source", "")
	obs.obs_data_set_default_string(settings, "label_default", "Alarm if in this state.")
	audio_default_settings(settings)
end

--
-- A function named script_update will be called when settings are changed
function script_update(settings)
	script_log("update")

	alarm_source = obs.obs_data_get_string(settings, "alarm_source")

	update_rule_settings(default_rule, settings)

	send_default_rule()
end

-- a function named script_load will be called on startup
function script_load(settings)
	script_log("script load")

	obs.obs_frontend_add_event_callback(frontend_event)

	bootstrap_rule_settings(default_rule, settings)

	examine_source_states()
	enum_sources(hook_source)

	local sh = obs.obs_get_signal_handler()
	-- signals received
	obs.signal_handler_connect(sh, "source_create", source_create)
	obs.signal_handler_connect(sh, "source_destroy", source_destroy)
	obs.signal_handler_connect(sh, "source_activate", source_activate)
	obs.signal_handler_connect(sh, "source_deactivate", source_deactivate)
	obs.signal_handler_add(sh, "void lua_mic_check_request_audio_sources()")
	obs.signal_handler_connect(sh, "lua_mic_check_request_audio_sources", request_audio_sources)
	obs.signal_handler_add(sh, "void lua_mic_check_source_rule(int id, string rule_json)")
	obs.signal_handler_connect(sh, "lua_mic_check_source_rule", update_source_rule)

	-- sent and received
	obs.signal_handler_add(sh, "void lua_mic_check_request_rules()")
	obs.signal_handler_connect(sh, "lua_mic_check_request_rules", request_rules)

	-- signals sent
	obs.signal_handler_add(sh, "void lua_mic_check_source_mute(ptr source)")
	obs.signal_handler_add(sh, "void lua_mic_check_default_rule(string rule_json)")
	obs.signal_handler_add(sh, "void lua_mic_check_video_source_status(string name, bool active, bool in_current_scene)")
	obs.signal_handler_add(sh, "void lua_mic_check_trigger_activate(int time, int timeout)")
	obs.signal_handler_add(sh, "void lua_mic_check_alarm_activate()")
	obs.signal_handler_add(sh, "void lua_mic_check_alarm_reset()")

	obs.signal_handler_signal(sh, "lua_mic_check_request_rules", nil)
	obs.signal_handler_signal(sh, "lua_mic_check_alarm_reset", nil)

	obs.timer_add(tick, sample_rate)
end

function script_unload()
	set_alarm_visible(false)
	-- these crash OBS
	--obs.timer_remove(update_frames)
end
